README.md
[![Build Status](https://travis-ci.org/vti/cpan-audit.svg?branch=master)](https://travis-ci.org/vti/cpan-audit) [![Build Status](https://img.shields.io/appveyor/ci/vti/cpan-audit/master.svg?logo=appveyor)](https://ci.appveyor.com/project/vti/cpan-audit/branch/master) [![MetaCPAN Release](https://badge.fury.io/pl/CPAN-Audit.svg)](https://metacpan.org/release/CPAN-Audit) [![Kritika Status](https://kritika.io/users/vti/repos/vti+cpan-audit/heads/master/status.svg)](https://kritika.io/users/vti/repos/vti+cpan-audit) [![Kritika Status](https://kritika.io/users/vti/repos/vti+cpan-audit/heads/master/status.svg?type=coverage)](https://kritika.io/users/vti/repos/vti+cpan-audit) [![Kritika Status](https://kritika.io/users/vti/repos/vti+cpan-audit/heads/master/status.svg?type=deps)](https://kritika.io/users/vti/repos/vti+cpan-audit)
# NAME

cpan-audit - Audit CPAN modules

# SYNOPSIS

cpan-audit \[command\] \[options...\]

Commands:

    module         [version range]    audit module with optional version range (all by default)
    dist|release   [version range]    audit distribution with optional version range (all by default)
    deps           [directory]        audit dependencies from the directory (. by default)
    installed                         audit all installed modules
    show           [advisory id]      show information about specific advisory

Options:

    --no-color    switch off colors
    --no-corelist ignore modules bundled with perl version
    --ascii       use ascii output
    --quiet       be quiet
    --verbose     be verbose
    --help|h      help message

Examples:

    cpan-audit dist Catalyst-Runtime
    cpan-audit dist Catalyst-Runtime 7.0
    cpan-audit dist Catalyst-Runtime >5.48

    cpan-audit module Catalyst 7.0

    cpan-audit deps .
    cpan-audit deps /path/to/distribution

    cpan-audit installed
    cpan-audit installed local/

    cpan-audit show CPANSA-Mojolicious-2018-03

# DESCRIPTION

`cpan-audit` is a command line application that checks the modules or distributions for known vulnerabilities. It is using
its internal database that is automatically generated from a hand-picked database
[https://github.com/vti/cpan-security-advisory](https://github.com/vti/cpan-security-advisory).

`cpan-audit` does not connect to anything, that is why it is important to keep it up to date. Every update of the internal
database is released as a new version.

`cpan-audit` can automatically detect dependencies from the following sources:

- `Carton`

    Parses `cpanfile.snapshot` file and checks the distribution versions.

- `cpanfile`

    Parses `cpanfile` taking into account the required versions.

It is assumed that if the required version of the module is less than a version of a release with a known vulnerability
fix, then the module is considered affected.

# LICENSE

Copyright (C) Viacheslav Tykhanovskyi.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension CPAN-Audit

{{$NEXT}}

0.15 2019-03-09T09:47:36Z

    - regenerate database fixing Plack-Middleware-Session distribution name

0.14 2019-01-26T10:23:21Z

    [ADVISORIES]
    CPANSA-Dancer2
    CPANSA-HTTP-Session2
    CPANSA-Plack-Middleware-Session-Cookie

0.13 2018-11-22T20:38:09Z

    - --no-corelist option by MCRayRay
    - test fixes

0.12 2018-11-11T19:43:25Z

    - require Module::CoreList latest version

0.11 2018-11-11T18:57:53Z

    - check core modules by James Raspass

0.10 2018-11-07T20:17:30Z

    - --quiet option
    - small refactoring
    - require the latest version of Pod::Usage

0.09 2018-11-05T21:17:35Z

    - do not hide db from pause (#7)

0.08 2018-10-17T18:10:41Z

    [ADVISORIES]
    - CPANSA-Net-DNS
    - CPANSA-PAR
    - CPANSA-PAR-Packer
    - CPANSA-RT-Authen-ExternalAuth
    - CPANSA-Tk
    - CPANSA-UI-Dialog (updated)
    - CPANSA-XML-LibXML

0.07 2018-10-16T21:37:20Z

    - test fixes

0.06 2018-10-16T19:19:22Z

    - use name instead of fullname
    - fix installed modules discovery

0.05 2018-10-15T19:36:39Z

    [ADVISORIES]
    - CPANSA-MHonArc
    - CPANSA-Module-Signature
    - CPANSA-libapreq2
    - CPANSA-mod_perl
    - CPANSA-Compress-Raw-Bzip2
    - CPANSA-Compress-Raw-Zlib

    [IMPROVEMENTS]
    - kritika.io and metacpan badges

0.04 2018-10-14T10:56:27Z

    [FEATURES]
    - install command accepts path to installations

    [IMPROVEMENTS]
    - get rid of Carton dependency
    - more test coverage
    - CI integrations
    - perl 5.8 compat

0.03 2018-10-13T12:59:36Z

    [ADVISORIES]
    - CPANSA-App-Github-Email
    - CPANSA-Crypt-OpenSSL-DSA
    - CPANSA-Crypt-Passwd-XS
    - CPANSA-DBD-MariaDB
    - CPANSA-Dancer
    - CPANSA-Data-Dumper
    - CPANSA-Email-Address
    - CPANSA-Encode
    - CPANSA-ExtUtils-MakeMaker
    - CPANSA-FCGI
    - CPANSA-Fake-Encode
    - CPANSA-Fake-Our
    - CPANSA-File-DataClass
    - CPANSA-File-Path
    - CPANSA-HTTP-Tiny
    - CPANSA-Imager
    - CPANSA-PathTools

    [FEATURES]
    - new installed command to audit all installed modules
    - cpan.snapshot support by Takumi Akiyama (github.com/akiym)

0.02 2018-10-09T08:24:36Z

    - support perl 5.8

0.01 2018-10-08T06:39:07Z

    - original version



t/lib/TestCommand.pm
package TestCommand;
use strict;
use warnings;
use Capture::Tiny qw(capture);

sub command {
    shift;
    my @args = @_;

    my ( $stdout, $stderr, $exit ) = capture {
        system $^X, 'script/cpan-audit', '--no-corelist', @args;
    };

    return ( $stdout, $stderr, $exit );
}

1;


t/data/installed/perl5/lib/perl5/Catalyst.pm
=pod

=head1 NAME

Catalyst

=cut

package      Catalyst  ;

our $VERSION = '5.0'; # No BumpVersion


lib/CPAN/Audit/Query.pm
package CPAN::Audit::Query;
use strict;
use warnings;
use CPAN::Audit::Version;

sub new {
    my $class = shift;
    my (%params) = @_;

    my $self = {};
    bless $self, $class;

    $self->{db} = $params{db} || {};

    return $self;
}

sub advisories_for {
    my $self = shift;
    my ( $distname, $version_range ) = @_;

    my $dist = $self->{db}->{dists}->{$distname};
    return unless $dist;

    my @advisories = @{ $dist->{advisories} };
    my @versions   = @{ $dist->{versions} };

    if ( !$version_range ) {
        return @advisories;
    }

    my $version_checker = CPAN::Audit::Version->new;

    my @all_versions = map { $_->{version} } @versions;
    my @selected_versions;

    foreach my $version (@all_versions) {
        if ( $version_checker->in_range( $version, $version_range ) ) {
            push @selected_versions, $version;
        }
    }

    if ( !@selected_versions ) {
        return;
    }

    my @matched_advisories;
    foreach my $advisory (@advisories) {
        my @affected_versions =
          $version_checker->affected_versions( \@all_versions,
            $advisory->{affected_versions} );
        next unless @affected_versions;

        foreach my $affected_version ( reverse @affected_versions ) {
            if ( $version_checker->in_range( $affected_version, $version_range )
              )
            {
                push @matched_advisories, $advisory;
                last;
            }
        }
    }

    if ( !@matched_advisories ) {
        return;
    }

    return @matched_advisories;
}

1;


lib/CPAN/Audit/Version.pm
package CPAN::Audit::Version;
use strict;
use warnings;
use version;

sub new {
    my $class = shift;

    my $self = {};
    bless $self, $class;

    return $self;
}

sub in_range {
    my $self = shift;
    my ( $version, $range ) = @_;

    return unless defined $version && defined $range;

    my @ands = split /\s*,\s*/, $range;

    return unless defined( $version = eval { version->parse($version) } );

    foreach my $and (@ands) {
        my ( $op, $range_version ) = $and =~ m/^(<=|<|>=|>|==|!=)?\s*([^\s]+)$/;

        return
          unless
          defined( $range_version = eval { version->parse($range_version) } );

        $op = '>=' unless defined $op;

        if ( $op eq '<' ) {
            return unless $version < $range_version;
        }
        elsif ( $op eq '<=' ) {
            return unless $version <= $range_version;
        }
        elsif ( $op eq '>' ) {
            return unless $version > $range_version;
        }
        elsif ( $op eq '>=' ) {
            return unless $version >= $range_version;
        }
        elsif ( $op eq '==' ) {
            return unless $version == $range_version;
        }
        elsif ( $op eq '!=' ) {
            return unless $version != $range_version;
        }
        else {
            return 0;
        }
    }

    return 1;
}

sub affected_versions {
    my $self = shift;
    my ($available_versions, $range) = @_;

    my @affected_versions;
    foreach my $version (@$available_versions) {
        if ($self->in_range($version, $range)) {
            push @affected_versions, $version;
        }
    }

    return @affected_versions;
}

1;


lib/CPAN/Audit/Installed.pm
package CPAN::Audit::Installed;
use strict;
use warnings;
use File::Find ();
use Cwd        ();

sub new {
    my $class = shift;
    my (%params) = @_;

    my $self = {};
    bless $self, $class;

    $self->{db} = $params{db};
    $self->{cb} = $params{cb};

    return $self;
}

sub find {
    my $self = shift;
    my (@inc) = @_;

    @inc = @INC unless @inc;
    @inc = grep { defined && -d $_ } map { Cwd::realpath($_) } @inc;

    my %seen;
    my @deps;

    File::Find::find(
        {
            wanted => sub {
                my $path = $File::Find::fullname;

                if ( $path && -f $path && m/\.pm$/ ) {
                    return unless my $module = module_from_file($path);

                    return unless my $distname = $self->{db}->{module2dist}->{$module};

                    my $dist = $self->{db}->{dists}->{$distname};
                    if ( $dist->{main_module} eq $module ) {
                        return if $seen{$module}++;

                        return unless my $version = module_version($path);

                        push @deps, { dist => $distname, version => $version };

                        if ( $self->{cb} ) {
                            $self->{cb}->(
                                {
                                    path     => $path,
                                    distname => $distname,
                                    version  => $version
                                }
                            );
                        }
                    }
                }
            },
            follow      => 1,
            follow_skip => 2,
        },
        @inc
    );

    return @deps;
}

# https://metacpan.org/source/ABELTJE/V-0.13/V.pm
sub module_version {
    my ($parsefile) = @_;

    open my $mod, '<', $parsefile or die $!;

    my $inpod = 0;
    my $result;
    local $_;
    while (<$mod>) {
        $inpod = /^=(?!cut)/ ? 1 : /^=cut/ ? 0 : $inpod;
        next if $inpod || /^\s*#/;

        chomp;
        next unless m/([\$*])(([\w\:\']*)\bVERSION)\b.*\=/;
        my $eval = qq{
            package CPAN::Audit::_version;
            no strict;

            local $1$2;
            \$$2=undef; do {
                $_
            }; \$$2
        };
        local $^W = 0;
        $result = eval($eval);
        warn "Could not eval '$eval' in $parsefile: $@" if $@;
        $result = "undef" unless defined $result;
        last;
    }
    close $mod;
    return $result;
}

sub module_from_file {
    my ($path) = @_;

    my $module;

    open my $fh, '<', $path or return;
    while ( my $line = <$fh> ) {
        if ( $line =~ m/package\s+(.*?)\s*;/ms ) {
            $module = $1;
            last;
        }
    }
    close $fh;

    return unless $module;
}

1;


lib/CPAN/Audit/Discover/Cpanfile.pm
package CPAN::Audit::Discover::Cpanfile;
use strict;
use warnings;
use Module::CPANfile;

sub new {
    my $class = shift;

    my $self = {};
    bless $self, $class;

    return $self;
}

sub discover {
    my $self = shift;
    my ($cpanfile_path) = @_;

    my $cpanfile = Module::CPANfile->load($cpanfile_path);

    my $prereqs = $cpanfile->prereqs->as_string_hash;

    my @deps;
    foreach my $phase ( keys %$prereqs ) {
        foreach my $type ( keys %{ $prereqs->{$phase} } ) {
            foreach my $module ( keys %{ $prereqs->{$phase}->{$type} } )
            {
                my $version = $prereqs->{$phase}->{$type}->{$module};

                next if $module eq 'perl';

                push @deps,
                  {
                    module       => $module,
                    version      => $version,
                    phase        => $phase,
                    relationship => $type,
                  };
            }
        }
    }

    return @deps;
}

1;


lib/CPAN/Audit.pm
package CPAN::Audit;
use 5.008001;
use strict;
use warnings;
use version;
use CPAN::Audit::Installed;
use CPAN::Audit::Discover;
use CPAN::Audit::Version;
use CPAN::Audit::Query;
use CPAN::Audit::DB;
use Module::CoreList;

our $VERSION = "0.14";

sub new {
    my $class = shift;
    my (%params) = @_;

    my $self = {};
    bless $self, $class;

    $self->{ascii}       = $params{ascii};
    $self->{verbose}     = $params{verbose};
    $self->{quiet}       = $params{quiet};
    $self->{no_color}    = $params{no_color};
    $self->{no_corelist} = $params{no_corelist};
    $self->{interactive} = $params{interactive};

    if ( !$self->{interactive} ) {
        $self->{ascii}    = 1;
        $self->{no_color} = 1;
    }

    $self->{db}       = CPAN::Audit::DB->db;
    $self->{query}    = CPAN::Audit::Query->new( db => $self->{db} );
    $self->{discover} = CPAN::Audit::Discover->new( db => $self->{db} );

    return $self;
}

sub command {
    my $self = shift;
    my ( $command, @args ) = @_;

    my %dists;

    if (!$self->{no_corelist}
        && (   $command eq 'dependencies'
            || $command eq 'deps'
            || $command eq 'installed' )
        )
    {
        # Find core modules for this perl version first.
        # This way explictly installed versions will overwrite.
        if ( my $core = $Module::CoreList::version{$]} ) {
            while ( my ( $mod, $ver ) = each %$core ) {
                my $dist = $self->{db}{module2dist}{$mod} or next;

                $dists{$dist} = $ver if version->parse($ver) > $dists{$dist};
            }
        }
    }

    if ( $command eq 'module' ) {
        my ( $module, $version_range ) = @args;
        $self->fatal("Usage: module <module> [version-range]") unless $module;

        my $distname = $self->{db}->{module2dist}->{$module};

        if ( !$distname ) {
            $self->message("__GREEN__Module '$module' is not in database");
            return 0;
        }

        $dists{$distname} = $version_range || '';
    }
    elsif ( $command eq 'release' || $command eq 'dist' ) {
        my ( $distname, $version_range ) = @args;
        $self->fatal("Usage: dist|release <module> [version-range]")
          unless $distname;

        if ( !$self->{db}->{dists}->{$distname} ) {
            $self->message("__GREEN__Distribution '$distname' is not in database");
            return 0;
        }

        $dists{$distname} = $version_range || '';
    }
    elsif ( $command eq 'show' ) {
        my ($advisory_id) = @args;
        $self->fatal("Usage: show <advisory-id>") unless $advisory_id;

        my ($release) = $advisory_id =~ m/^CPANSA-(.*?)-(\d+)-(\d+)$/;
        $self->fatal("Invalid advisory id") unless $release;

        my $dist = $self->{db}->{dists}->{$release};
        $self->fatal("Unknown advisory id") unless $dist;

        my ($advisory) =
          grep { $_->{id} eq $advisory_id } @{ $dist->{advisories} };
        $self->fatal("Unknown advisory id") unless $advisory;

        $self->print_advisory($advisory);

        return 0;
    }
    elsif ( $command eq 'dependencies' || $command eq 'deps' ) {
        my ($path) = @args;
        $path = '.' unless defined $path;

        $self->fatal("Usage: deps <path>") unless -d $path;

        my @deps = $self->{discover}->discover($path);

        $self->message( 'Discovered %d dependencies', scalar(@deps) );

        foreach my $dep (@deps) {
            my $dist = $dep->{dist}
              || $self->{db}->{module2dist}->{ $dep->{module} };
            next unless $dist;

            $dists{$dist} = $dep->{version};
        }
    }
    elsif ( $command eq 'installed' ) {
        $self->message_info('Collecting all installed modules. This can take a while...');

        my @deps = CPAN::Audit::Installed->new(
            db => $self->{db},
            $self->{verbose}
            ? (
                cb => sub {
                    my ($info) = @_;

                    $self->message( '%s: %s-%s', $info->{path}, $info->{distname}, $info->{version} );
                }
              )
            : ()
        )->find(@ARGV);

        foreach my $dep (@deps) {
            my $dist = $dep->{dist}
              || $self->{db}->{module2dist}->{ $dep->{module} };
            next unless $dist;

            $dists{ $dep->{dist} } = $dep->{version};
        }
    }
    else {
        $self->fatal("Error: unknown command: $command. See -h");
    }

    my $total_advisories = 0;

    if (%dists) {
        my $query = $self->{query};

        foreach my $distname ( sort keys %dists ) {
            my $version_range = $dists{$distname};

            my @advisories = $query->advisories_for( $distname, $version_range );

            $version_range = 'Any'
              if $version_range eq '' || $version_range eq '0';

            if (@advisories) {
                $self->message( '__RED__%s (requires %s) has %d advisories__RESET__',
                    $distname, $version_range, scalar(@advisories) );

                foreach my $advisory (@advisories) {
                    $self->print_advisory($advisory);
                }
            }

            $total_advisories += @advisories;
        }
    }

    if ($total_advisories) {
        $self->message( '__RED__Total advisories found: %d__RESET__', $total_advisories );

        return $total_advisories;
    }
    else {
        $self->message_info('__GREEN__No advisories found__RESET__');

        return 0;
    }
}

sub message_info {
    my $self = shift;

    return if $self->{quiet};

    $self->message(@_);
}

sub message {
    my $self = shift;

    $self->_print( *STDOUT, @_ );
}

sub fatal {
    my $self = shift;
    my ( $msg, @args ) = @_;

    $self->_print( *STDERR, "Error: $msg", @args );
    exit 255;
}

sub print_advisory {
    my $self = shift;
    my ($advisory) = @_;

    $self->message("  __BOLD__* $advisory->{id}");

    print "    $advisory->{description}\n";

    if ( $advisory->{affected_versions} ) {
        print "    Affected range: $advisory->{affected_versions}\n";
    }

    if ( $advisory->{fixed_versions} ) {
        print "    Fixed range: $advisory->{fixed_versions}\n";
    }

    if ( $advisory->{cves} ) {
        print "\n    CVEs: ";
        print join ', ', @{ $advisory->{cves} };
        print "\n";
    }

    if ( $advisory->{references} ) {
        print "\n    References:\n";
        foreach my $reference ( @{ $advisory->{references} || [] } ) {
            print "    $reference\n";
        }
    }

    print "\n";
}

sub _print {
    my $self = shift;
    my ( $fh, $format, @params ) = @_;

    my $msg = @params ? ( sprintf( $format, @params ) ) : ($format);

    if ( $self->{no_color} ) {
        $msg =~ s{__BOLD__}{}g;
        $msg =~ s{__GREEN__}{}g;
        $msg =~ s{__RED__}{}g;
        $msg =~ s{__RESET__}{}g;
    }
    else {
        $msg =~ s{__BOLD__}{\e[39;1m}g;
        $msg =~ s{__GREEN__}{\e[32m}g;
        $msg =~ s{__RED__}{\e[31m}g;
        $msg =~ s{__RESET__}{\e[0m}g;

        $msg .= "\e[0m";
    }

    print $fh "$msg\n";
}

1;
__END__

=encoding utf-8

=head1 NAME

CPAN::Audit - Audit CPAN distributions for known vulnerabilities

=head1 SYNOPSIS

    use CPAN::Audit;

=head1 DESCRIPTION

CPAN::Audit is a module and a database at the same time. It is used by L<cpan-audit> command line application to query
for vulnerabilities.

=head1 LICENSE

Copyright (C) Viacheslav Tykhanovskyi.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Viacheslav Tykhanovskyi E<lt>viacheslav.t@gmail.comE<gt>

=head1 CREDITS

Takumi Akiyama (github.com/akiym)

James Raspass (github.com/JRaspass)

MCRayRay (github.com/MCRayRay)

=cut
